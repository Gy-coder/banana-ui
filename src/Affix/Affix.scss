.g-affix {
  width: 100%;
  &-wrapper {
    width: 100%;
    height: 100%;
    z-index: 200;
  }
}
