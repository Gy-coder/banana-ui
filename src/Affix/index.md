---
title: Affix
nav:
  title: Affix
  path: /components
---

# Affix

将页面元素钉在可视范围。

# 何时使用

当内容区域比较长，需要滚动页面时，这部分内容对应的操作或者导航需要在滚动范围内始终展现。常用于侧边菜单和按钮组合。

页面可视范围过小时，慎用此功能以免遮挡页面内容。

# 代码演示

## 基础使用

<code src="./demos/basic.tsx" />

## API

| 参数 | 说明 | 类型 | 是否必须 | 默认值 |
| ---- | ---- | ---- | -------- | ------ |
