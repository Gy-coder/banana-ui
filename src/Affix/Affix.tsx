import React, { CSSProperties, useEffect, useRef } from 'react';
import classnames from 'classnames';
import './Affix.scss';

export interface AffixProps {
  distance?: number;
  className?: string;
  style?: CSSProperties;
}

const Affix: React.FC<AffixProps> = (props) => {
  const { distance, className, style, children } = props;
  const top = useRef<number>(0);
  const wrapperRef = useRef<HTMLDivElement>(null);
  const affixRef = useRef<HTMLDivElement>(null);
  useEffect(() => {
    window.addEventListener('scroll', handleScroll);
    changeTop();
    return () => {
      window.removeEventListener('scroll', handleScroll);
    };
  }, []);
  const changeTop = () => {
    if (window.scrollY === 0) {
      const value = affixRef.current!.getBoundingClientRect().top;
      top.current = value;
    } else {
      const { scrollX, scrollY } = window;
      window.scrollTo(scrollX, 0);
      top.current = affixRef.current!.getBoundingClientRect().top;
      window.scrollTo(scrollX, scrollY);
    }
  };
  const handleScroll = () => {
    if (window.scrollY > top.current - distance!) {
      const { top, left, right, bottom } =
        wrapperRef.current!.getBoundingClientRect();
      wrapperRef.current!.style.width = right - left + 'px';
      wrapperRef.current!.style.height = bottom - top + 'px';
      wrapperRef.current!.style.left = left + 'px';
      wrapperRef.current!.style.top = distance + 'px';
      wrapperRef.current!.style.position = 'fixed';
    } else {
      wrapperRef.current!.style.position = 'static';
    }
  };
  return (
    <div className="g-affix" ref={affixRef}>
      <div
        className={classnames('g-affix-wrapper', className)}
        ref={wrapperRef}
        style={style}
      >
        {children}
      </div>
    </div>
  );
};

Affix.defaultProps = {
  distance: 0,
};

export default Affix;
