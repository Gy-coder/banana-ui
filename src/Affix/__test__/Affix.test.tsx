import { fireEvent, render, RenderResult } from '@testing-library/react';
import React from 'react';
import '@testing-library/jest-dom';
import Affix from '../Affix';

let wrapper: RenderResult;
describe('test affix', () => {
  beforeEach(() => {
    wrapper = render(
      <Affix distance={20}>
        <div>Affix</div>
      </Affix>,
    );
  });
  it('Affix shoule render correct', () => {
    expect(wrapper.queryByText('Affix')).toBeInTheDocument();
    expect(wrapper.asFragment()).toMatchSnapshot();
  });
  it('Affix should have position fixed', () => {
    fireEvent.scroll(window, { target: { screenY: 1000 } });
    expect(
      wrapper.container.querySelectorAll('.g-affix-wrapper')[0],
    ).toHaveStyle('position: fixed');
  });
});
