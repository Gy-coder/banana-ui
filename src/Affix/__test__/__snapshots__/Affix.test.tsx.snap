// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test affix Affix shoule render correct 1`] = `
<DocumentFragment>
  <div
    class="g-affix"
  >
    <div
      class="g-affix-wrapper"
    >
      <div>
        Affix
      </div>
    </div>
  </div>
</DocumentFragment>
`;
